File: audit-trail-sdk-go/integration_test.go
//go:build integration
// +build integration

package audittrail

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestIntegration(t *testing.T) {
	client, err := NewClientBuilder().
		ServerURL("http://localhost:8080").
		Build()
	assert.NoError(t, err)

	ctx := context.Background()

	t.Run("Log event", func(t *testing.T) {
		event := NewEvent(
			NewUserActor("go-sdk-test", "Go SDK Tester"),
			CreateAction("SDK Go integration test"),
			DocumentResource("doc-go-001", "Test Document"),
			NewEventMetadata("sdk-go-test", "tenant-001"),
		)

		response, err := client.Log(ctx, event)

		assert.NoError(t, err)
		assert.NotEmpty(t, response.ID)
		assert.NotEmpty(t, response.Hash)
		assert.Equal(t, "STORED", response.Status)
	})

	t.Run("Log batch events", func(t *testing.T) {
		events := []Event{
			NewEvent(
				NewUserActor("user-1", "User 1"),
				CreateAction("Batch test 1"),
				DocumentResource("doc-1", "Doc 1"),
				NewEventMetadata("test", "tenant-001"),
			),
			NewEvent(
				NewUserActor("user-2", "User 2"),
				UpdateAction("Batch test 2"),
				DocumentResource("doc-2", "Doc 2"),
				NewEventMetadata("test", "tenant-001"),
			),
		}

		response, err := client.LogBatch(ctx, events)

		assert.NoError(t, err)
		assert.Equal(t, 2, response.Total)
		assert.Equal(t, 2, response.Succeeded)
	})

	t.Run("Get by ID", func(t *testing.T) {
		// Create event first
		event := NewEvent(
			NewUserActor("test", "Test"),
			ReadAction("Get test"),
			DocumentResource("doc-get", "Doc"),
			NewEventMetadata("test", "tenant-001"),
		)
		created, err := client.Log(ctx, event)
		assert.NoError(t, err)

		// Retrieve it
		retrieved, err := client.GetByID(ctx, created.ID)

		assert.NoError(t, err)
		assert.NotNil(t, retrieved)
		assert.Equal(t, created.ID, retrieved.ID)
	})

	t.Run("Get non-existent returns nil", func(t *testing.T) {
		result, err := client.GetByID(ctx, "non-existent-id")

		assert.NoError(t, err)
		assert.Nil(t, result)
	})

	t.Run("Search events", func(t *testing.T) {
		criteria := SearchCriteria{
			TenantID: "tenant-001",
			Page:     0,
			Size:     10,
		}

		result, err := client.Search(ctx, criteria)

		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.NotNil(t, result.Items)
	})

	t.Run("Quick search", func(t *testing.T) {
		result, err := client.QuickSearch(ctx, "document", "tenant-001", 0, 10)

		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.NotNil(t, result.Items)
	})
}

File: audit-trail-sdk-js/test/integration.test.ts
import { describe, it, expect } from 'vitest';
import {
  AuditTrailClient,
  ActorBuilder,
  ActionBuilder,
  ResourceBuilder,
  MetadataBuilder,
  EventBuilder
} from '../src';

// Ces tests nécessitent le serveur en cours d'exécution
// Pour exécuter: npm run test:integration
describe.skip('Integration Tests', () => {
  const client = AuditTrailClient.builder()
    .serverUrl('http://localhost:8080')
    .build();

  it('should log event successfully', async () => {
    const event = new EventBuilder()
      .actor(ActorBuilder.user('sdk-js-test', 'JS SDK Tester'))
      .action(ActionBuilder.create('SDK JS integration test'))
      .resource(ResourceBuilder.document('doc-js-001', 'Test Document'))
      .metadata(new MetadataBuilder()
        .source('sdk-js-test')
        .tenantId('tenant-001')
        .build())
      .build();

    const response = await client.log(event);

    expect(response.id).toBeDefined();
    expect(response.hash).toBeDefined();
    expect(response.status).toBe('STORED');
  });

  it('should log batch events successfully', async () => {
    const events = [
      new EventBuilder()
        .actor(ActorBuilder.user('batch-user-1', 'User 1'))
        .action(ActionBuilder.create('Batch event 1'))
        .resource(ResourceBuilder.document('batch-doc-1', 'Doc 1'))
        .metadata(new MetadataBuilder().source('batch-test').tenantId('tenant-001').build())
        .build(),
      new EventBuilder()
        .actor(ActorBuilder.user('batch-user-2', 'User 2'))
        .action(ActionBuilder.update('Batch event 2'))
        .resource(ResourceBuilder.document('batch-doc-2', 'Doc 2'))
        .metadata(new MetadataBuilder().source('batch-test').tenantId('tenant-001').build())
        .build(),
    ];

    const response = await client.logBatch(events);

    expect(response.total).toBe(2);
    expect(response.succeeded).toBe(2);
    expect(response.failed).toBe(0);
  });

  it('should get event by id', async () => {
    // First create an event
    const event = new EventBuilder()
      .actor(ActorBuilder.user('test-user', 'Test'))
      .action(ActionBuilder.read('Read test'))
      .resource(ResourceBuilder.document('doc-1', 'Doc'))
      .metadata(new MetadataBuilder().source('test').tenantId('tenant-001').build())
      .build();

    const created = await client.log(event);
    const retrieved = await client.getById(created.id);

    expect(retrieved).not.toBeNull();
    expect(retrieved?.id).toBe(created.id);
  });

  it('should return null for non-existent event', async () => {
    const result = await client.getById('non-existent-id-12345');
    expect(result).toBeNull();
  });

  it('should search events', async () => {
    // First create an event to search
    const event = new EventBuilder()
      .actor(ActorBuilder.user('search-test-user', 'Search Tester'))
      .action(ActionBuilder.create('Searchable event'))
      .resource(ResourceBuilder.document('search-doc-001', 'Search Doc'))
      .metadata(new MetadataBuilder()
        .source('search-test')
        .tenantId('tenant-001')
        .build())
      .build();

    await client.log(event);

    // Wait a bit for indexing
    await new Promise(resolve => setTimeout(resolve, 1000));

    const result = await client.search({
      tenantId: 'tenant-001',
      actorId: 'search-test-user',
      page: 0,
      size: 10
    });

    expect(result.items).toBeDefined();
    expect(Array.isArray(result.items)).toBe(true);
  });

  it('should quick search events', async () => {
    const result = await client.quickSearch('document', 'tenant-001', 0, 10);

    expect(result.items).toBeDefined();
    expect(Array.isArray(result.items)).toBe(true);
  });
});
